Fixing 1950 for wrong incrementation of changes count in admin dashboard

HugsDaniel opened the pull request:
Fixing 1950 for wrong incrementation of changes count in admin dashboard, displaying the same message no matter the number of changes made

#### What? Why?

Closes #1950 

As suggested by @lin-d-hop I simply removed the counter for unsaved changes thus displaying the same status message no matter how many unsaved changes have been made.

#### What should we test?

statusMessages display correctly for unsaved changes made in customers pane in admin dashboard

enricostano commented on app/assets/javascripts/admin/index_utils/services/pending_changes.js.coffee:
```diff
@@ -6,7 +6,7 @@ angular.module("admin.indexUtils").factory "pendingChanges", ($q, resources, Sta
       @pendingChanges["#{id}"] = {} unless @pendingChanges.hasOwnProperty("#{id}")
       @pendingChanges["#{id}"]["#{attr}"] = change
-      StatusMessage.display('notice', "You have made #{@changeCount(@pendingChanges)} unsaved changes")
+      StatusMessage.display('notice', "You have unsaved changes")
```
I know that I'm asking for an unrelated thing here... but if we could take advantage of this change to create a i18n key and use `t()` instead of a literal string, that would be awesome. But only if it takes really small added effort.

In any case, many thanks for your contribution @HugsDaniel 🎉

HugsDaniel commented:
I added i18n filter and replaced switch with if/else statement as suggested. 

I have a question though : it seems that the status message display "You have unsaved changes" is done in pending_changes.js.coffee by the "add" method. 
The if/else statement in bulk_product_update.js.coffee doesn't seem to be fulfilling this particular task. That's where the issue was coming from : 
'         when 0 then StatusMessage.clear()
-        when 1 then StatusMessage.display 'notice', t("one_product_unsaved")
-        else StatusMessage.display 'notice', t("products_unsaved", n: count)'
These lines would have displayed the status message correctly in the beginning if displayDirtyProducts was in charge of the task, but it doesn't seem to be the case.

So here is my question : what is the purpose of displayDirtyProducts ? Is it unrelated to this issue and so should not be edited ?

HugsDaniel commented:
Ok so I got confused with Angular controllers, bulk_product_update.js.coffee is actually for product update, which is none of our concern here considering we're in the customers controller... Sorry about that !

enricostano commented:
Thanks @HugsDaniel ! Do we have any test covering this behavior? Would be too much to ask you to add some test to cover it? Many thanks for your time!

HugsDaniel commented:
I'd be happy to do it. I've got some syntax error messages when I run the tests though, would you know where it's coming from ?
```
An error occurred while loading ./spec/javascripts/unit/admin/index_utils/services/pending_changes_spec.js.coffee.
Failure/Error: load_dependency(file) { result = super }

SyntaxError:
  /home/hugs/code/HugsDaniel/openfoodnetwork/spec/javascripts/unit/admin/index_utils/services/pending_changes_spec.js.coffee:1: syntax error, unexpected '\n', expecting keyword_do_LAMBDA or tLAMBEG
  /home/hugs/code/HugsDaniel/openfoodnetwork/spec/javascripts/unit/admin/index_utils/services/pending_changes_spec.js.coffee:4: syntax error, unexpected '\n', expecting keyword_do_LAMBDA or tLAMBEG
  /home/hugs/code/HugsDaniel/openfoodnetwork/spec/javascripts/unit/admin/index_utils/services/pending_changes_spec.js.coffee:7: syntax error, unexpected ->, expecting end-of-input

```

I've got : 
bundler 1.16.1
rspec-core 3.7.0
active-support 3.2.21

sauloperez commented:
@HugsDaniel how did you run that file? were you running the JS unit tests? what exact commands did you use?

HugsDaniel commented:
Ok I tried with rspec... now I do it with zeus and Karma. Sorry to waste your time like that, I'm all set now thanks !

sauloperez commented:
Yes, I figured. It looked like it was ruby who was trying to interpret that coffee script.

HugsDaniel commented:
Sorry for the 100000+ commits, I got confused with my branches. 
I added a test on customers_spec.rb to test the status message display. It passes on my machine, can you test it as well, just to make sure ? Thx

sauloperez commented on app/assets/javascripts/admin/index_utils/services/pending_changes.js.coffee:
```diff
@@ -6,7 +6,7 @@ angular.module("admin.indexUtils").factory "pendingChanges", ($q, resources, Sta
     add: (id, attr, change) =>
       @pendingChanges["#{id}"] = {} unless @pendingChanges.hasOwnProperty("#{id}")
       @pendingChanges["#{id}"]["#{attr}"] = change
```
Is there a chance to simplify this algorithm that checks for pending changes? Now, if I'm not mistaken, it's enough with `@pendingChanges` being a boolean to know whether there are changes or not, or even no `@pendingChanges` and simply displaying the message whenever a change is applied.

HugsDaniel replied:
I'm not 100% sure about this (I'm just beginning with Angular as a matter of fact, I'm more comfortable with Rails) but this method doesn't check for pending changes, it adds the pending changes in client side memory (again I'm not quite sure but at least it adds them in the DOM), and then wait for submitAll(form) method to save them.

Could someone with Angular knowledge confirm this ? Or maybe shed some light on the add method ?

sauloperez replied:
perhaps @mkllnk can help us here? I understand that @HugsDaniel so up to you. We might have to create an issue out of this though.

Matt-Yorkley replied:
So the idea is just to display a "you have unsaved changes" message, without the count? That seems like a reasonable solution.

I wouldn't change `@pendingChanges`. It's not a boolean, it's a hash that holds the form changes before they are saved.

mkllnk replied:
Yes, a boolean wouldn't know which changes have been applied and which ones are still on the way.

sauloperez replied:
:+1:

mkllnk commented:
@HugsDaniel I noticed that you submitted 10 commits for two changed lines. It is very good to make many commits during development. But when submitting a pull request, we usually clean up our history with `git rebase -i upstream/master`. Reviewers don't need to follow your detours. In this case, it doesn't matter so much, because we can just look at all changes at once. But if a pull request becomes bigger, it's much easier to go through the commits one by one (if they are well written).

HugsDaniel commented:
I just cleaned the unnecessary commit as suggested by @mkllnk

myriamboure commented:
Wahou, it seems we have another PR to queue for staging in order to test @RachL !!! Awesome @HugsDaniel thank you for your contributions, you reeeally help us !!!

mkllnk commented:
Staged on https://staging1.openfood.com.au/.

RachL commented:
Awesome ! I will test this tonight @myriamboure @HugsDaniel @mkllnk

myriamboure commented:
Your proactivity gives me joy @RachL thanks a lot! You rock :-)

myriamboure commented:
Very quick thing to test, here are testing notes: https://docs.google.com/document/d/1S3SyqatnieS-buQcSvGHykCRTVhxDPKLDFhDHSmAUq4/edit?usp=sharing 
I just noticed accessing the back office was super slow from France but maybe that's because my internet connexion is slow. I guess it can't be related to that PR but wanted to mention it in case.
Ready to go for me.

Next one is for you @RachL I was afraid Maikel and the Aus team need the staging when they wake up :-)

mkllnk commented:
Thanks Myriam. The slowness can't be this code change. I will merge this.
